ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/task/AbstractRetryingSwitchingPeerTask.java

```diff
@@ -0,0 +1,153 @@
+    triedPeers.add(peer);
+  }
+
+  protected abstract CompletableFuture<T> executeTaskOnCurrentPeer(final EthPeer peer);
+
+  @Override
+  protected CompletableFuture<T> executePeerTask(final Optional<EthPeer> assignedPeer) {
+
+    final Optional<EthPeer> maybePeer =
+        assignedPeer
+            .filter(u -> getRetryCount() == 1) // first try with the assigned peer if present
+            .map(Optional::of)
+            .orElseGet(this::selectNextPeer); // otherwise select a new one from the pool
+
+    if (maybePeer.isEmpty()) {
+      traceLambda(
+          LOG,
+          "No peer found to try to execute task at attempt {}, tried peers {}",
+          this::getRetryCount,
+          triedPeers::toString);
+      final var ex = new NoAvailablePeersException();
+      return CompletableFuture.failedFuture(ex);
+    }
+
+    final EthPeer peerToUse = maybePeer.get();
+    assignPeer(peerToUse);
+
+    traceLambda(
+        LOG,
+        "Trying to execute task on peer {}, attempt {}",
+        this::getAssignedPeer,
+        this::getRetryCount);
+
+    return executeTaskOnCurrentPeer(peerToUse)
+        .thenApply(
+            peerResult -> {
+              traceLambda(
+                  LOG,
+                  "Got result {} from peer {}, attempt {}",
+                  peerResult::toString,
+                  peerToUse::toString,
+                  this::getRetryCount);
+              result.complete(peerResult);
+              return peerResult;
+            });
+  }
+
+  @Override
+  protected void handleTaskError(final Throwable error) {
+    if (isPeerFailure(error)) {
+      getAssignedPeer().ifPresent(peer -> failedPeers.add(peer));
+    }
+    super.handleTaskError(error);
+  }
+
+  @Override
+  protected boolean isRetryableError(final Throwable error) {
+    return error instanceof TimeoutException || isPeerFailure(error);
+  }
+
+  private Optional<EthPeer> selectNextPeer() {
+    final Optional<EthPeer> maybeNextPeer = remainingPeersToTry().findFirst();
+
+    if (maybeNextPeer.isEmpty()) {
+      // tried all the peers, restart from the best one but excluding the failed ones
+      refreshPeers();
+      triedPeers.retainAll(failedPeers);
+      return remainingPeersToTry().findFirst();
+    }
+
+    return maybeNextPeer;
+  }
+
+  private Stream<EthPeer> remainingPeersToTry() {
+    return getEthContext()
+        .getEthPeers()
+        .streamBestPeers()
+        .filter(peer -> !peer.isDisconnected() && !triedPeers.contains(peer));
+  }
+
+  private void refreshPeers() {
+    final EthPeers peers = getEthContext().getEthPeers();
+    // If we are at max connections, then refresh peers disconnecting one of the failed peers,
+    // or the least useful
+    if (peers.peerCount() >= peers.getMaxPeers()) {
+      failedPeers.stream()
+          .filter(peer -> !peer.isDisconnected())
+          .findAny()
+          .or(() -> peers.streamAvailablePeers().sorted(peers.getBestChainComparator()).findFirst())
```

Comment:
it is interesting that getBestChainComparator should give the worst peer first

Reply:
yes I thought the same :)